### PR TITLE
[13.0][MIG] sale_invoice_plan

### PR DESCRIPTION
**Pull request by kittiu**
Standard migration

- [x] Merge changes in #967 to this PR
- [x] included this fixed on v12 https://github.com/OCA/sale-workflow/pull/1130

### CODE REVIEW

File: sale_invoice_plan/models/sale.py
Comment (sudhir-erpharbor) [inline sale_invoice_plan/models/sale.py]: Rename round variable to something else as this is a built-in method of python.

Travis: sale_invoice_plan/models/sale.py:7: [W0622(redefined-builtin), ] Redefining built-in 'round'

File: sale_invoice_plan/wizard/sale_make_invoice_advance.py
Comment (sudhir-erpharbor) [inline sale_invoice_plan/wizard/sale_make_invoice_advance.py]: It should be "invoice_move_ids" instead of "invoice_ids"

Test case error:
AttributeError: 'sale.invoice.plan' object has no attribute 'invoice_ids'

File: sale_invoice_plan/tests/test_sale_invoice_plan.py
Comment (sudhir-erpharbor) [inline sale_invoice_plan/tests/test_sale_invoice_plan.py:45]: invoice_move_ids

File: sale_invoice_plan/tests/test_sale_invoice_plan.py
Comment (sudhir-erpharbor) [inline sale_invoice_plan/tests/test_sale_invoice_plan.py:71]: invoice_move_ids

File: sale_invoice_plan/tests/test_sale_invoice_plan.py
Comment (sudhir-erpharbor) [inline sale_invoice_plan/tests/test_sale_invoice_plan.py:113]: invoice_move_ids

File: sale_invoice_plan/tests/test_sale_invoice_plan.py
Comment (sudhir-erpharbor) [inline sale_invoice_plan/tests/test_sale_invoice_plan.py:170]: invoice_move_ids

File: sale_invoice_plan/models/sale.py
Comment (sudhir-erpharbor) [inline sale_invoice_plan/models/sale.py]: I think it should be account.move since invoice object is removed.

File: sale_invoice_plan/README.rst
Comment (sudhir-erpharbor) [inline sale_invoice_plan/README.rst]: 13.0 in the url

File: sale_invoice_plan/README.rst
Comment (sudhir-erpharbor) [inline sale_invoice_plan/README.rst]: 13.0

**Review by sudhir-erpharbor — COMMENTED** [review]
Important changes

File: sale_invoice_plan/tests/test_sale_invoice_plan.py
Comment (kittiu) [reply sale_invoice_plan/tests/test_sale_invoice_plan.py:170]: invoice_ids is available for sale.order

**Comment by kittiu** [comment]
Ready fo review.

**Comment by sudhir-erpharbor** [comment]
Please fix travis.

sale_invoice_plan/__init__.py:1: [C7902(missing-readme), ] Missing ./README.rst file. Template here: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: sale_invoice_plan/models/sale.py
Comment (sudhir-erpharbor) [inline sale_invoice_plan/models/sale.py]: delivered. There is a space between "deliver" & "ed"

Comment (kittiu) [reply sale_invoice_plan/models/sale.py]: Thank you, fixed.

File: sale_invoice_plan/README.rst
Comment (kittiu) [reply sale_invoice_plan/README.rst]: I removed README.rst, it will be created auto by ocabot when merged.

**Comment by kittiu** [comment]
The error in on this line,
`sale_invoice_plan/wizard/sale_make_planned_invoice.py:3:1: F401 'odoo.api' imported but unused
`
Fixed.

File: sale_invoice_plan/tests/test_sale_invoice_plan.py
Comment (kittiu) [reply sale_invoice_plan/tests/test_sale_invoice_plan.py:71]: invoice_ids is already correct.

**Comment by kittiu** [comment]
@sudhir-erpharbor thank you for your reviewed, all fixed.
Thanks!

**Comment by kittiu** [comment]
The remaining problem is on runbot, which I think it might be runbot config and not related to this module.
```
2019-10-08 06:27:53,633 150 [1;33m[1;49mWARNING[0m openerp_template odoo.addons.phone_validation.tools.phone_validation: The `phonenumbers` Python module is not installed, contact numbers will not be verified. Please install the `phonenumbers` Python module. 
```

**Comment by sudhir-erpharbor** [comment]
Code review LGTM

**Comment by kittiu** [comment]
> Code review LGTM

:+1: please approve if you see appropriate. Thank you for your reviews.

**Comment by kittiu** [comment]
@Saran440 please can you help review this.

File: sale_invoice_plan/models/sale.py
Comment (bealdav) [inline sale_invoice_plan/models/sale.py]: s/SaleOder/SaleOrder

Comment (kittiu) [reply sale_invoice_plan/models/sale.py]: @bealdav thanks for reviewing. I am not sure what your comment is about :p

**Comment by kittiu** [comment]
I had misc enhancement in [12.0] https://github.com/OCA/sale-workflow/pull/967
I will wait until that one is merged, and will add the same change here.

File: sale_invoice_plan/models/sale.py
Comment (bealdav) [reply sale_invoice_plan/models/sale.py]: oh, it's just a typo error.
You wrote `SaleOder`, I suppose you would write `SaleOrder` :)

**Comment by Saran440** [comment]
@kittiu Functional Test : Found 2 bugs
**1.Invoice Plan calculate only first, when post invoice**
step :
- Create sales and invoice plan 4 installment
- Confirm and receive order 1/4
Selection_003
- Create Invoice by Plan -> Create Next Invoice
Selection_004
- Post invoice
Selection_006
-  Sale invoice plan will back to first
Selection_005

**2.Due Date and Invoice Date in Invoices mistake**
step :
- Create normal process and use invoice plan
- Click create all remaining invoices
Selection_007
- Invoice date should be depend plan date and due date shouldn't same as.
Selection_008

**Comment by newtratip** [comment]
Red CI

**Comment by AaronHForgeFlow** [comment]
If I select and advance invoice int he plan:
image

I cannot see this button until i send some quantity, I think this is not correct:
image

**Comment by AaronHForgeFlow** [comment]
Also, if I select advance payment in the invoice plan the first invoice I generate using plan invoice will be generate correctly, but not the second one. See the example below:

I select advance invoice for 50% of the total and then the next invoice is for another 40%:
image

The first invoice is created correctly, however the second one is not, it generates again another down payment for the 50%, which is not correct:

image

**Comment by kittiu** [comment]
@AaronHForgeFlow thank you for your review. I will check that out.

**Comment by AaronHForgeFlow** [comment]
Thank you @kittiu  Perhaps it is that I am not understanding correctly the concept of advance payment. Imagine I have this situation where I invoice in advance 50% of the total and then 40% the next month:
image

The advance invoice is created correctly for half price of the sales order. However, when creating the next invoice it creates an invoice line for the 40% (correct) but it is including a credit note discount for the 40% of the advance (see image below):
image

I was expecting just the first two invoice lines to be created, but not that extra line. Maybe am I missing something here, if you could shed some light here it would be awesome. Thanks for the work.

**Review by AaronHForgeFlow — COMMENTED** [review]
I tested now the module without choosing the "Advance On 1st" Option. I selected 3 Installments. I created the first invoice based on plan and I posted it. However, the invoice plan was not updated:
image

Additionally, if I create another invoice from the plan, then it works, like it were the first one, and I get something like this:
image

Sorry for all the separate reviews :)

**Comment by kittiu** [comment]
@AaronHForgeFlow thank you so much for the detailed test. I have made quite a number of fixes. Can you test again please?

The concept is this,
image

Also, as you say here,
> The advance invoice is created correctly for half price of the sales order. However, when creating the next invoice it creates an invoice line for the 40% (correct) but it is including a credit note discount for the 40% of the advance (see image below):

I think the standard Odoo advance also works this way, having the minus value of its part. And I just try to reuse it.

**Comment by kittiu** [comment]
> If I select and advance invoice int he plan:
> image
> 
> I cannot see this button until i send some quantity, I think this is not correct:
> image

Just this one, I guess this should be bug since v12 also, did you have chance to test it?

**Comment by AaronHForgeFlow** [comment]
> 
> Just this one, I guess this should be bug since v12 also, did you have chance to test it?

Yes, and in v12 it was happening the same thing. Thanks for the quick update. :)

**Comment by AaronHForgeFlow** [comment]
> I think the standard Odoo advance also works this way, having the minus value of its part. And I just try to reuse it.

Yes. Please disregard that comment.

**Review by AaronHForgeFlow — APPROVED** [review]
Tested on local. All issues are solved now. :+1:

**Comment by kittiu** [comment]
@Saran440 your issue is also fixed. Would you approve?

**Comment by kittiu** [comment]
Not really sure why runbot is failed, there is no error.

**Comment by AaronHForgeFlow** [comment]
@kittiu 

I think it is this warning:
`2020-05-01 16:10:12,740 165 WARNING openerp_test odoo.addons.base.models.decimal_precision: Deprecated call to decimal_precision.get_precision(<application>), use digits=<application> instead`

**Comment by kittiu** [comment]
@AaronHForgeFlow :+1:

**Comment by kittiu** [comment]
@AaronHForgeFlow can merge?

**Comment by kittiu** [comment]
@AaronHForgeFlow @HviorForgeFlow 
Trying to clear all the sale/purchase invoice plan up to v13.

**Comment by AaronHForgeFlow** [comment]
For me it is ok to move forward.  cc @JordiBForgeFlow

**Comment by kittiu** [comment]
@OCA/crm-sales-marketing-maintainers can we merge this?

**Review by Saran440 — APPROVED** [review]
Code Reviewed.

you should seperate 2 commits
- [IMP] $MODULE: black, isort, prettier for pre-commit pattern
- [MIG] xxx for easy reviewed

**Comment by kittiu** [comment]
> Code Reviewed.
> 
> you should seperate 2 commits
> 
> * [IMP] $MODULE: black, isort, prettier for pre-commit pattern
> * [MIG] xxx for easy reviewed

Done

**Review by HviorForgeFlow — CHANGES_REQUESTED** [review]
Functional Review:

Using first month in Advance when confirming Quotation:
image

Trying to fix that I decided to remove advance installment to move forward. I notice when you remove an installment Create Invoice Plan Menu button disappears. I've tried to solve it removing Invoice Plan and recreating it once Sale Order is confirm, same behavior I can't create the invoice plan. So regarding there is no specified group security, everyone can modify the invoice plan model, I've got stuck went trying to fix. Same happens if I edit Invoice Plan before Confirm the Order.

Does Invoice Plan should be immutable once created? For example, I want to create a non linear invoice plan reusing a wizard creation's one and modify percentages.

@kittiu

**Comment by kittiu** [comment]
@HviorForgeFlow thank you for your review.

I can't produce the behavior you mentioned. What kind of product you use, is it product or service? If you use product, only if product delivered, then the "Create Invoice Plan" button will be shown.

I try with both product (make sure it delivered) and service. And you can remove or recreate invoice plan once it is confirmed. This is intentional. Only when some invoice(s) is created, it will be immutable.

**Comment by HviorForgeFlow** [comment]
Ok. When Invoice Policy is invoice what is delivered it seems to work fine, not mention in readme, I had the problems with the other policy. I'm still having problem when use Advance on 1st invoice. Error in previous image I uploaded.

**Comment by kittiu** [comment]
Error because you didnt input the percentage for advance yet.

**Comment by kittiu** [comment]
I am not sure I understand your problem yet. ;) may ne you can elaborate a bit.

**Comment by HviorForgeFlow** [comment]
sale_invoice_plan_error

**Comment by kittiu** [comment]
That is because the advance is 0 percent. You should edit it first. For example make it 10 percent.

**Review by HviorForgeFlow — APPROVED** [review]

**Comment by rousseldenis** [comment]
@kittiu Is this ready ?

**Comment by kittiu** [comment]
Yes, very much. :)

**Comment by kittiu** [comment]
@rousseldenis, can merged? :)

**Comment by kittiu** [comment]
> @rousseldenis, can merged? :)
@rousseldenis, would you?

File: sale_invoice_plan/views/sale_view.xml
Comment (rousseldenis) [inline sale_invoice_plan/views/sale_view.xml:87]: @kittiu Could you make a fix for this. This generates a warning on view load. 

See https://travis-ci.com/github/OCA/sale-workflow/jobs/494482747#L1767

Thanks !

**Pull request merged**
